File: solution/368. Largest Divisible Subset/368_03.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        
        int max_len = 0;
        int index = 0;
        vector<int> dp(n, 1);
        vector<int> prev(n, -1);
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                        prev[i] = j;
                    }
                }
            }
            if (dp[i] > max_len) {
                index = i;
                max_len = dp[i];
            }
        }
        
        vector<int> res;
        while (index != -1) {
            res.push_back(nums[index]);
            index = prev[index];
        }
        return res;
    }
};

File: solution/368. Largest Divisible Subset/368_03.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        nums.sort()
        l = 0
        idx = 0
        dp = [0] * n
        prev = [-1] * n
        for i in range(n):
            for j in range(0, i):
                if nums[i] % nums[j] == 0 and dp[j] + 1 > dp[i]:
                    dp[i] = dp[j] + 1
                    prev[i] = j
            if dp[i] > l:
                l = dp[i]
                idx = i
        
        res = []
        while idx != -1:
            res.append(nums[idx])
            idx = prev[idx]
        return res
    
